File: cineva/packages/audio_engine/native/cineva_dsp/src/drc.c
#include "drc.h"

#include <math.h>
#include <string.h>

#include "dsp_util.h"

void cv_drc_init(cv_drc *drc, double sample_rate) {
  memset(drc, 0, sizeof(*drc));
  drc->sample_rate = cv_clamp_d(sample_rate, 8000.0, 192000.0);
  drc->threshold_db = -24.0;
  drc->ratio = 2.5;
  drc->knee_db = 6.0;
  drc->attack_ms = 15.0;
  drc->release_ms = 250.0;
  drc->makeup_db = 0.0;
  drc->mix = 1.0;
  drc->detector = 1;
  drc->env_db = -70.0;
  drc->gain_db = 0.0;
  cv_drc_set(drc, drc->enable, drc->threshold_db, drc->ratio, drc->knee_db,
             drc->attack_ms, drc->release_ms, drc->makeup_db, drc->mix * 100.0,
             drc->detector);
}

void cv_drc_set(cv_drc *drc, int enable, double threshold_db, double ratio,
                double knee_db, double attack_ms, double release_ms,
                double makeup_db, double mix_percent, int detector) {
  drc->enable = enable ? 1 : 0;
  drc->threshold_db = cv_clamp_d(cv_sanitize_d(threshold_db), -60.0, 0.0);
  drc->ratio = cv_clamp_d(cv_sanitize_d(ratio), 1.0, 20.0);
  drc->knee_db = cv_clamp_d(cv_sanitize_d(knee_db), 0.0, 24.0);
  drc->attack_ms = cv_clamp_d(cv_sanitize_d(attack_ms), 0.5, 200.0);
  drc->release_ms = cv_clamp_d(cv_sanitize_d(release_ms), 20.0, 1000.0);
  drc->makeup_db = cv_clamp_d(cv_sanitize_d(makeup_db), -6.0, 12.0);
  drc->mix = cv_clamp_d(cv_sanitize_d(mix_percent), 0.0, 100.0) / 100.0;
  drc->detector = detector == 0 ? 0 : 1;
  drc->att_coef = cv_onepole_coef(drc->sample_rate, drc->attack_ms / 1000.0);
  drc->rel_coef = cv_onepole_coef(drc->sample_rate, drc->release_ms / 1000.0);
  drc->rms_coef = cv_onepole_coef(drc->sample_rate, 0.010);
}

void cv_drc_reset(cv_drc *drc) {
  drc->env_db = -70.0;
  drc->gain_db = 0.0;
}

/* Gain de compression (dB, <= 0) pour un niveau détecté (dB). */
static double cv_drc_compute_gain_db(const cv_drc *drc, double level_db) {
  double thr = drc->threshold_db;
  double knee = drc->knee_db;
  double slope = 1.0 / drc->ratio - 1.0; /* négatif */
  double over = level_db - thr;
  if (knee > 0.0 && 2.0 * over < knee && 2.0 * over > -knee) {
    /* Soft knee quadratique. */
    double x = over + knee * 0.5;
    double gr = slope * x * x / (2.0 * knee);
    return gr;
  }
  if (over <= 0.0) return 0.0;
  return slope * over;
}

void cv_drc_process(cv_drc *drc, float *const *io, int channels, int frames) {
  if (!drc->enable) return;
  if (channels < 1 || channels > 2) return;
  double rms_state = drc->rms_state_cache;
  double makeup = cv_db_to_lin(drc->makeup_db);
  double dry_mix = 1.0 - drc->mix;

  for (int i = 0; i < frames; i++) {
    double l = cv_sanitize_f(io[0][i]);
    double r = channels > 1 ? cv_sanitize_f(io[1][i]) : l;

    /* Détection linkée (un seul détecteur pour les deux canaux). */
    double level;
    if (drc->detector == 0) {
      double pk = fabs(l) > fabs(r) ? fabs(l) : fabs(r);
      level = pk;
    } else {
      double ms = (l * l + r * r) * 0.5;
      rms_state = drc->rms_coef * rms_state + (1.0 - drc->rms_coef) * ms;
      level = sqrt(rms_state);
    }
    double level_db = cv_lin_to_db(level);

    /* Enveloppe : montée rapide (attack), descente lente (release). */
    if (level_db > drc->env_db) {
      drc->env_db = drc->att_coef * drc->env_db + (1.0 - drc->att_coef) * level_db;
    } else {
      drc->env_db = drc->rel_coef * drc->env_db + (1.0 - drc->rel_coef) * level_db;
    }

    double target_gain = cv_drc_compute_gain_db(drc, drc->env_db);
    /* Lissage supplémentaire du gain (anti-pompage). */
    if (target_gain < drc->gain_db) {
      drc->gain_db = drc->att_coef * drc->gain_db + (1.0 - drc->att_coef) * target_gain;
    } else {
      drc->gain_db = drc->rel_coef * drc->gain_db + (1.0 - drc->rel_coef) * target_gain;
    }
    double wet_gain = cv_db_to_lin(drc->gain_db) * makeup;
    double gl = dry_mix + wet_gain * drc->mix;
    io[0][i] = (float)(l * gl);
    if (channels > 1) io[1][i] = (float)(r * gl);
  }
  drc->rms_state_cache = rms_state;
}

double cv_drc_gain_db(const cv_drc *drc) { return drc->gain_db; }

File: cineva/packages/audio_engine/lib/src/dsp/drc.dart
/// Compresseur soft-knee stéréo-linké — miroir de `drc.c`.
library;

import 'dart:math' as math;

import 'biquad.dart' as dsp;
import 'dart:typed_data';

class DynamicRangeCompressor {
  DynamicRangeCompressor(this.sampleRate) {
    set(false, -24, 2.5, 6, 15, 250, 0, 100, 1);
  }

  final double sampleRate;
  bool enable = false;
  double thresholdDb = -24;
  double ratio = 2.5;
  double kneeDb = 6;
  double attackMs = 15;
  double releaseMs = 250;
  double makeupDb = 0;
  double mix = 1;
  int detector = 1;

  double attCoef = 0;
  double relCoef = 0;
  double rmsCoef = 0;
  double envDb = -70;
  double gainDb = 0;
  double rmsStateCache = 0;

  void set(bool enable, double thresholdDb, double ratio, double kneeDb,
      double attackMs, double releaseMs, double makeupDb, double mixPercent,
      int detector) {
    this.enable = enable;
    this.thresholdDb = dsp.clampD(dsp.sanitizeD(thresholdDb), -60, 0);
    this.ratio = dsp.clampD(dsp.sanitizeD(ratio), 1, 20);
    this.kneeDb = dsp.clampD(dsp.sanitizeD(kneeDb), 0, 24);
    this.attackMs = dsp.clampD(dsp.sanitizeD(attackMs), 0.5, 200);
    this.releaseMs = dsp.clampD(dsp.sanitizeD(releaseMs), 20, 1000);
    this.makeupDb = dsp.clampD(dsp.sanitizeD(makeupDb), -6, 12);
    mix = dsp.clampD(dsp.sanitizeD(mixPercent), 0, 100) / 100;
    this.detector = detector == 0 ? 0 : 1;
    attCoef = dsp.onepoleCoef(sampleRate, this.attackMs / 1000);
    relCoef = dsp.onepoleCoef(sampleRate, this.releaseMs / 1000);
    rmsCoef = dsp.onepoleCoef(sampleRate, 0.010);
  }

  void reset() {
    envDb = -70;
    gainDb = 0;
    rmsStateCache = 0;
  }

  double _computeGainDb(double levelDb) {
    final double thr = thresholdDb;
    final double knee = kneeDb;
    final double slope = 1 / ratio - 1;
    final double over = levelDb - thr;
    if (knee > 0 && 2 * over < knee && 2 * over > -knee) {
      final double x = over + knee * 0.5;
      return slope * x * x / (2 * knee);
    }
    if (over <= 0) return 0;
    return slope * over;
  }

  void process(List<Float32List> io, int channels, int frames) {
    if (!enable) return;
    if (channels < 1 || channels > 2) return;
    double rmsState = rmsStateCache;
    final double makeup = dsp.dbToLin(makeupDb);
    final double dryMix = 1 - mix;

    for (int i = 0; i < frames; i++) {
      final double l = dsp.sanitizeD(io[0][i]);
      final double r = channels > 1 ? dsp.sanitizeD(io[1][i]) : l;

      double level;
      if (detector == 0) {
        final double al = l.abs();
        final double ar = r.abs();
        level = al > ar ? al : ar;
      } else {
        final double ms = (l * l + r * r) * 0.5;
        rmsState = rmsCoef * rmsState + (1 - rmsCoef) * ms;
        level = rmsState <= 0 ? 0 : math.sqrt(rmsState);
      }
      final double levelDb = dsp.linToDb(level);

      if (levelDb > envDb) {
        envDb = attCoef * envDb + (1 - attCoef) * levelDb;
      } else {
        envDb = relCoef * envDb + (1 - relCoef) * levelDb;
      }

      final double targetGain = _computeGainDb(envDb);
      if (targetGain < gainDb) {
        gainDb = attCoef * gainDb + (1 - attCoef) * targetGain;
      } else {
        gainDb = relCoef * gainDb + (1 - relCoef) * targetGain;
      }
      final double wetGain = dsp.dbToLin(gainDb) * makeup;
      final double g = dryMix + wetGain * mix;
      io[0][i] = l * g;
      if (channels > 1) io[1][i] = r * g;
    }
    rmsStateCache = rmsState;
  }
}
